### DISCUSSION TITLE
Is there a way to force/hack zero 'contribution' from holiday binary variables?

### DISCUSSION THREAD

**Discussion by gedkitchener-CT** [facebookexperimental/Robyn#329] [Q&A] [answered]
In its current form, all holidays are lumped together when calling Prophet.

But since the effect of different holidays can be different, I have decided to leave out "holidays" from Prophet

`prophet_vars = c("trend", "season", "weekday")              # leave out "holiday"`

and instead generate a number of binary flags to represent certain holidays individually:

`context_vars = c(..., "Christmas_Day", "New_Years_Day")`
`factor_vars = c(..., "Christmas_Day", "New_Years_Day")`

I then run the model, and in the `prophet_decomp` output, I'll get something like this for `Christmas_Day`:

<img width="361" alt="image" src="https://user-images.githubusercontent.com/53869388/155993767-ade3f7b9-5841-4b95-8cad-63a65124fdd4.png">

While this may make sense from a purely statistical point of view, it doesn't from a business/interpretation point of view. That is to say, if we look at a non-'Christmas' day, there _should not_ be any 'Christmas' contribution - the contribution (whether positive or negative) should only be defined on the 'Christmas' day, and should be zero on all other days (by definition). i.e., it should look more like this:

<img width="362" alt="image" src="https://user-images.githubusercontent.com/53869388/155993804-9b54dcad-8258-41de-8071-a5bfcb5e6d1e.png">

Here, the 'Christmas' variable has a combined contribution of `-£1,000` (compare this with first example, where it had a combined contribution of `+£11,000`). Incidentally, that difference of `£12,000` is being 'artificially' taken away from other variables, making the rest of the output less reflective of reality.

Is there a way to force/hack zero 'contribution' from holiday binary variables? (when the binary flag = 0)

**Comment by kyletgoldberg** [comment]
Great question - while their isn't a way to force zero contribution currently, there is a way to force the coefficient for a particular variable to have a negative (or positive) sign, which may fix the problem. Could you try setting `context_signs = c('negative')` for this variable in question and seeing whether that solves the problem?

**Comment by gedkitchener-CT** [comment]
Yeah, I get what you mean, but I'm always a bit hesitant of enforcing the coefficient to have a negative or positive sign.

Take Christmas day for example - its effect may be negative (because people are with their families) but it may very well be positive (because people have free time to browse e-commerce sites). Both are entirely plausible, and so I'd want to avoid explciitly setting the coefficient to be positive/negative because that will essentially ingrain my biased/incomplete view of the world onto the model.

**Comment by gedkitchener-CT** [comment]
Thinking about it more, I may treat it as an organic (free) media variable, i.e., when the media variable is zero, its contribution to the dependent variable can only be zero. There is the issue that the model will attempt to apply adstock to the variable, but if I fix the adstock decay rate as zero, then that should solve that problem. Given it's an organic (read: free) variable, I don't have to worry about dealing with the Hill transformation.

In summary, here's my proposed solution:
 * treat it as a media variable so that zero 'impressions' will give zero return: `organic_vars = c("Christmas_Day")`
 * allow flexibility on whether its contribution can be positive or negative: `organic_signs = c("default")`
 * to ensure no adstock is applied, I'll set `Christmas_Day_thetas = 0` in the `hyperparameters` (~`alpha` and `gamma` aren't applicable because it's an organic variable~)

In my head, that solves the problem, but let me know if you think I'm missing something. I'll report back with the results.

**Comment by kyletgoldberg** [comment]
I think that organic media hack may end up working the same as your initial example where all the non-christmas days were positive and christmas was zero. My recommendation of using the `context_signs` variable to force the sign would definitely need to be taken with caution and only used with prior experience showing good reason for that sign to be a certain direction. Additionally, if you are planning to go back and re-run the model you can fix the signs after the initial run (where the sign would be unconstrained) so that the initial problem you highlighted should be solved for in the subsequent run.

**Reply by gedkitchener-CT** [reply]
After re-running the model using my hack, I seem to be getting the 'behaviour' I want, i.e., zero contribution on days when `x=0` and non-zero contribution on days when `x=1`.

The only reason I decided to use `organic_vars` to hack my problem is because the regression takes the form `(β1 * x1) + ...`, and so whenever `x1=0`, that term can only be zero (ignoring adstock effects), and whenever `x1=1`, it would have a non-zero value.

To be honest, I'm not exactly sure how binary variables (defined in `factor_vars`) were being treated in the Prophet decomposition, but I certainly wasn't getting the 'behaviour' I wanted.

**Comment by gedkitchener-CT** [comment, marked as answer]
**Solution:** treat the binary variable as an organic media variable.

**In practice, this means:**

- adding the variable name to `organic_vars` so that zero 'impressions' will give zero return, e.g., `organic_vars = c("Christmas_Day")`
- allow flexibility on whether the binary variable's contribution can be positive or negative, i.e., `organic_signs = c("default")`
- to ensure no adstock is applied, set suitable hyperparameters, i.e., `Christmas_Day_thetas = 0` in `hyperparameters`

After re-running the model using this 'hack', I seem to be getting the 'behaviour' I want, i.e., zero contribution on days when `x=0` and non-zero contribution on days when `x=1`.

**Reply by kyletgoldberg** [reply]
Thanks for the back and forth here and glad you found a solution that works for your need!

**Reply by swapnilpatil022** [reply]
@gedkitchener-CT Just a quick question, I am also having similar scenario like yours. In my case I am using Weibull PDF as a adstock function and this is what I have done to set the adstock effect to ~0, but still it is giving me contribution when holiday flag is zero. 

     hyperparameters <- list()
     for (ms in robyn_param_mmm_omni["paid_media_spends"]) {
       hyperparameters[[ paste(ms,"alphas",sep="_") ]] <- c(0.5, 3)
       hyperparameters[[ paste(ms,"gammas",sep="_") ]] <- c(0.1, 1)
       hyperparameters[[ paste(ms,"scales",sep="_") ]] <- c(0, 0.1)
       hyperparameters[[ paste(ms,"shapes",sep="_") ]] <- c(0, 10)
    }

    for (ms in robyn_param_mmm_omni["organic_vars"]) {
      hyperparameters[[ paste(ms,"alphas",sep="_") ]] <- c(0.001, 0.001)
      hyperparameters[[ paste(ms,"gammas",sep="_") ]] <- c(0.001, 0.001)
      hyperparameters[[ paste(ms,"scales",sep="_") ]] <- c(0.001, 0.001)
      hyperparameters[[ paste(ms,"shapes",sep="_") ]] <- c(0.001, 0.001)
     }
